discovery-provider/src/tasks/user_replica_set.py

```diff
@@ -0,0 +1,259 @@
+                if event_type == user_replica_set_manager_event_types_lookup['update_replica_set']:
+                    primary = args._primary
+                    secondaries = args._secondaries
+                    user_record = user_replica_set_events_lookup[user_id]["user"]
+                    user_record.updated_at = datetime.utcfromtimestamp(block_timestamp)
+                    user_record.primary = primary
+                    user_record.secondaries = secondaries
+
+                    # Update cnode endpoint string reconstructed from sp ID
+                    creator_node_endpoint_str = get_endpoint_string_from_sp_ids(
+                        self,
+                        update_task,
+                        primary,
+                        secondaries,
+                        redis
+                    )
+                    user_record.creator_node_endpoint = creator_node_endpoint_str 
+                    user_replica_set_events_lookup[user_id]["user"] = user_record
+                    user_replica_set_events_lookup[user_id]["events"].append(event_type)
+                # Process L2 Content Node operations
+                elif event_type == user_replica_set_manager_event_types_lookup['add_or_update_content_node']:
+                    cnode_record = parse_poa_cnode_record(
+                        self,
+                        update_task,
+                        session,
+                        entry,
+                        cnode_events_lookup[cnode_id]["content_node"]
+                    )
+                    if cnode_record is not None:
+                        cnode_events_lookup[cnode_id]["content_node"] = cnode_record
+                        cnode_events_lookup[cnode_id]["events"].append(event_type)
+            num_total_changes += len(user_events_tx)
+
+    # for each record in user_replica_set_events_lookup, invalidate the old record and add the new record
+    # we do this after all processing has completed so the user record is atomic by block, not tx
+    for user_id, value_obj in user_replica_set_events_lookup.items():
+        logger.info(f"user_replica_set.py | Replica Set Processing Adding {value_obj['user']}")
+        invalidate_old_user(session, user_id)
+        session.add(value_obj["user"])
+
+    for content_node_id, value_obj in cnode_events_lookup.items():
+        logger.info(f"user_replica_set.py | Content Node Processing Adding {value_obj['content_node']}")
+        invalidate_old_cnode_record(session, content_node_id)
+        session.add(value_obj["content_node"])
+
+    return num_total_changes, user_ids
+
+# Reconstruct endpoint string from primary and secondary IDs
+# Note that this is BEST EFFORT and may fail, however unlikely
+# In the case of failure where a given user has valid primary/secondaries but no 
+# endpoint string, a client must fetch the endpoint associated with ID 
+# Indexing CANNOT block on this endpoint string optimization.
+def get_endpoint_string_from_sp_ids(
```

Comment:
cc @SidSethi @dmanjunath I want to discuss a potential retry pattern here but non-blocking functionally

Reply:
yeah lets talk about this